applications/common/backend/src/main/java/pl/ds/kyanite/common/components/models/table/TableCellComponent.java

```diff
@@ -16,39 +16,135 @@
 
   @Inject
-  @Getter
-  @Default(intValues = 1)
-  private int rows;
+  private Integer rows;
+
+  @SlingObject
+  private Resource currentResource;
 
   @PostConstruct
   private void init() {
-    if (text.isEmpty()) {
+    if (StringUtils.isEmpty(text)) {
       text = DEFAULT_TEXT;
     }
   }
 
+  public int getColspan() {
+    return columns != null ? columns : computeColspan();
+  }
+
+  public int getRowspan() {
+    return rows != null ? rows : DEFAULT_ROWSPAN;
+  }
+
+  private int computeColspan() {
+    Optional<TableComponent> parentTable = Optional
+            .ofNullable(currentResource.getParent()) // table row
+            .map(Resource::getParent) // table head or body
+            .map(Resource::getParent) // the table
+            .map(tableResource -> tableResource.adaptTo(TableComponent.class));
+
+    if (parentTable.isEmpty() || parentTable.get().isScrollable()) {
+      return DEFAULT_COLSPAN;
+    }
+
+    Resource parentRow = currentResource.getParent();
+
+    List<TableCellData> tableCells = StreamSupport
+            .stream(parentRow.getChildren().spliterator(), false)
+            .map(cell -> new TableCellData(
+                    cell.getValueMap().get("columns", Integer.class),
+                    cell.getPath().equals(currentResource.getPath())))
+            .toList();
+
+    TableRowData tableRowData = new TableRowData(tableCells);
+
+    return computeColspan(tableRowData);
+  }
+
+  /**
+   * Computes colspan for a cell that hasn't got the columns property specified.
+   * Attempts to give each such cell the same colspan, if possible.
+   * It takes into account colspans from other cells in the same row.
+   */
+  private static int computeColspan(TableRowData row) {
+    int summaryAvailableColspan = TableComponent.DEFAULT_SUMMARY_COLSPAN
+            - row.getSummaryColspanTakenByCellsWithColumnsSpecified();
+
+    if (summaryAvailableColspan < 1) {
+      return 1;
+    }
+
+    int colspanAvailableForSingleCell = summaryAvailableColspan
+            / row.getNumberOfCellsWithColumnsUnspecified();
+    if (colspanAvailableForSingleCell == 0) {
+      colspanAvailableForSingleCell = 1;
+    }
+
+    if (summaryAvailableColspan
+            % row.getNumberOfCellsWithColumnsUnspecified() == 0) {
+      // all cells are able to receive the same colspan
+      return colspanAvailableForSingleCell;
+    }
+
+    // ... otherwise - we have some additional colspan space to distribute between cells
+    int colspanLeftToDistribute = summaryAvailableColspan
+            - row.getNumberOfCellsWithColumnsUnspecified() * colspanAvailableForSingleCell;
+
+    int indexOfCurrentCellInCellsWithColumnsUnspecified =
+            getIndexOfCurrentCellInCellsWithColumnsUnspecified(row.getTableCellData());
+
+    if (indexOfCurrentCellInCellsWithColumnsUnspecified < colspanLeftToDistribute) {
+      // when giving additional +1 to colspan, prefer cells from the left
+      return colspanAvailableForSingleCell + 1;
```

Comment:
We "give" the extra point to the first cell from the left (out of the three cells that we compute colspan for).